protocol/vm/vm.go

```diff
@@ -42,8 +42,12 @@ var TraceOut io.Writer
 
 func Verify(context *Context) (err error) {
 	defer func() {
-		if panErr := recover(); panErr != nil {
-			err = ErrUnexpected
+		if r := recover(); r != nil {
+			if rErr, ok := r.(error); ok {
+				err = errors.Sub(ErrUnexpected, rErr)
+			} else {
+				err = ErrUnexpected
```

Comment:
Since we're trying to preserve information in this PR, maybe also consider on this line:

```
err = errors.Wrap(ErrUnexpected, r)
```